src/stream/stream/cycle.rs

```diff
@@ -0,0 +1,69 @@
+use std::pin::Pin;
+
+use pin_project_lite::pin_project;
+
+use crate::stream::Stream;
+use crate::task::{Context, Poll};
+
+pin_project! {
+    /// A stream that will repeatedly yield the same list of elements
+    pub struct Cycle<S, T> {
+        #[pin]
+        source: S,
+        index: usize,
+        buffer: Vec<T>,
+        state: CycleState,
+    }
+}
+
+#[derive(Eq, PartialEq)]
+enum CycleState {
+    FromStream,
+    FromBuffer,
+}
+
+impl<T: Clone, S: Stream<Item = T>> Cycle<S, T> {
```

Comment:
I think this code is better.  I don't feel the need to derive a generic type T.
Do you think?

```rust
impl<S> Cycle<S, S::Item>
where
    S: Stream,
    S::Item: Clone,
{
```

Reply:
Is it also possible to use the same trick for the struct definition?